src/logging/ulog.c

```diff
@@ -0,0 +1,333 @@
+    }
+
+    ret = fs_write(&log->file, key, key_len);
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, val, 4);
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    return ULOG_SUCCESS;
+}
+
+void ULOG_StartDataPhase(ULOG_Inst_Type *log) { log->phase = ULOG_PHASE_DATA; }
+
+ULOG_Error_Type ULOG_LogString(ULOG_Inst_Type *log, const char *string,
+                               const size_t len,
+                               const ULOG_Log_Level_Type level) {
+    if (log == NULL || string == NULL) {
+        return ULOG_INVALID_PARAM;
+    }
+
+    if (log->phase != ULOG_PHASE_DATA) {
+        return ULOG_WRONG_PHASE;
+    }
+
+    Message_Header_Type header = {
+        .msg_type = 'L',
+        .msg_size = sizeof(uint8_t) + sizeof(uint64_t) + len,
+    };
+
+    int ret = fs_write(&log->file, &header, sizeof(header));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    const uint8_t log_level = level;
+    ret = fs_write(&log->file, &log_level, sizeof(log_level));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    const uint64_t uptime_us = k_uptime_get() * 1000;
+    ret = fs_write(&log->file, &uptime_us, sizeof(uptime_us));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, string, len);
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    return ULOG_SUCCESS;
+}
+
+ULOG_Error_Type ULOG_LogTaggedString(ULOG_Inst_Type *log, const char *string,
+                                     const size_t len,
+                                     const ULOG_Log_Level_Type level,
+                                     const uint16_t tag) {
+    if (log == NULL || string == NULL) {
+        return ULOG_INVALID_PARAM;
+    }
+
+    if (log->phase != ULOG_PHASE_DATA) {
+        return ULOG_WRONG_PHASE;
+    }
+
+    Message_Header_Type header = {
+        .msg_type = 'C',
+        .msg_size = sizeof(uint8_t) + sizeof(uint16_t) + sizeof(uint64_t) + len,
+    };
+
+    int ret = fs_write(&log->file, &header, sizeof(header));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    const uint8_t log_level = level;
+    ret = fs_write(&log->file, &log_level, sizeof(log_level));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, &tag, sizeof(tag));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    const uint64_t uptime_us = k_uptime_get() * 1000;
+    ret = fs_write(&log->file, &uptime_us, sizeof(uptime_us));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, string, len);
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    return ULOG_SUCCESS;
+}
+
+ULOG_Error_Type ULOG_Sync(ULOG_Inst_Type *log) {
+    if (log == NULL) {
+        return ULOG_INVALID_PARAM;
+    }
+
+    int ret = WriteSync(&log->file);
```

Comment:
Phase guard should be added for ULOG_Sync message, function shouldn't allow user to write Sync out of Data Phase of the ULog

Reply:
Good find, thanks!